x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSort.java

```diff
@@ -0,0 +1,263 @@
+                )
+            );
+            case DOUBLE -> new MvSort.EvaluatorFactory(
+                toEvaluator.apply(field),
+                toEvaluator.apply(order),
+                (blockFactory, fieldBlock, ascending) -> new MultivalueDedupeDouble((DoubleBlock) fieldBlock).sortToBlock(
+                    blockFactory,
+                    ascending
+                )
+            );
+            case NULL -> EvalOperator.CONSTANT_NULL_FACTORY;
+            default -> throw new IllegalArgumentException("unsupported type [" + field.dataType() + "]");
+        };
+    }
+
+    private record EvaluatorFactory(
+        EvalOperator.ExpressionEvaluator.Factory field,
+        EvalOperator.ExpressionEvaluator.Factory order,
+        TriFunction<BlockFactory, Block, Boolean, Block> sort
+    ) implements EvalOperator.ExpressionEvaluator.Factory {
+        @Override
+        public EvalOperator.ExpressionEvaluator get(DriverContext context) {
+            return new MvSort.Evaluator(context.blockFactory(), field.get(context), order.get(context), sort);
+        }
+
+        @Override
+        public String toString() {
+            return "MvSort[field=" + field + ", order=" + order + "]";
+        }
+    }
+
+    private static class Evaluator implements EvalOperator.ExpressionEvaluator {
+        private final BlockFactory blockFactory;
+        private final EvalOperator.ExpressionEvaluator field;
+        private final EvalOperator.ExpressionEvaluator order;
+        private final TriFunction<BlockFactory, Block, Boolean, Block> sort;
+
+        protected Evaluator(
+            BlockFactory blockFactory,
+            EvalOperator.ExpressionEvaluator field,
+            EvalOperator.ExpressionEvaluator order,
+            TriFunction<BlockFactory, Block, Boolean, Block> sort
+        ) {
+            this.blockFactory = blockFactory;
+            this.field = field;
+            this.order = order;
+            this.sort = sort;
+        }
+
+        @Override
+        public Block eval(Page page) {
+            try (Block fieldBlock = field.eval(page)) {
+                try (BytesRefBlock orderBlock = (BytesRefBlock) order.eval(page)) {
+                    boolean ascending = orderBlock.areAllValuesNull();  // if all are nulls, default is ascending
+                    if (ascending == false) { // the first not null value decides the order
+                        for (int p = 0; p < orderBlock.getPositionCount(); p++) {
+                            int count = orderBlock.getValueCount(p);
+                            if (count > 0) {
+                                BytesRef orderScratch = new BytesRef();
+                                BytesRef sortOrder = orderBlock.getBytesRef(orderBlock.getFirstValueIndex(p), orderScratch);
+                                ascending = sortOrder.utf8ToString().equalsIgnoreCase("DESC") == false;
+                                break;
+                            }
```

Comment:
It looks like you're looking for the first value in the block and locking the sort to that. I think we have two choices and this is probably wrong for either choice.

1. We lock the `ASC`/`DESC` choice to a constant - in which case I think we should use `fold` to get the value and assert at type resolution time that the field is `foldable`.
2. We don't lock the `ASC`/`DESC` choice to a constant, in which case we do something like pass a `BooleanVector` to the `sortToBlock` method.
  a. We could totally specialize when the parameter *is* a constant block `bock.asVector().isConstant()` and pass a constant down. That's going to be very common.

I think I'd prefer we do option 1 and force a constant and fold. We can always make the thing non-constant later.

Reply:
done.